pipeline/loading.py

```diff
@@ -56,6 +70,19 @@ def make_yaml_error_more_helpful(
 
 
 def parse_yaml_file(data: str | Path, filename: str | None = None) -> dict[str, Any]:
+    # If a fast parser is availabe and can parse the input without error then we just
+    # return it. This results in a very significant speed-up for large files. If there
+    # are errors then we re-parse using the pure Python parser which gives much more
+    # helpful error messages.
+    #
+    # Note that this _is_ covered by tests in CI but, because we're not combining
+    # coverage across multiple runs, we have to mark it as uncovered.
+    if FAST_PARSER is not None:  # pragma: no cover
+        try:
+            return FAST_PARSER.load(data)  # type: ignore[no-any-return]
+        except Exception:
+            pass
```

Comment:
One potential problem that makes me nervous is that there's a potential for an invisible bug here that leads to silent performance degradation, which is that if the fast parser errors but the slow parser doesn't, we end up silently falling back to the slow parser without saying.

I think it'd be better if we made sure to error every time the fast parser errors, with as helpful an exception message as we can manage under the circumstances.

Reply:
I know where you're coming from, but I think in this specific case I disagree. For users running things locally the slow path is the default anyway. For those that have a fast parser installed I want the experience to be purely an improvement and not introduce new failure modes.

When this library is used in Job Server it _would_ be helpful to know if the fast parser failed where the slow succeeded – again ideally without blowing up. We could probably log something here and then add a handler to Job Server which picks these up and pushes to Sentry so we get alerted.

I'm not sure this is the most effective use of engineering effort at the moment though. Beyond asking the user to rewrite their YAML there's not much we can _do_ about this parse failure; it's not like we've got a stack of alternative parsers we can switch to. And this is all intended as a short-term (ha!) hack until we improve the [installation story](https://github.com/opensafely-core/opensafely-cli/issues/277) to the point where we can start having hard dependencies on compiled libs. So overall, I take your point but it's not high up my list of worries at the moment.

Reply:
Fair enough! A log seems reasonable to me, but so does not treating this as high priority. YAML and silent performance degradation both just make me nervous.

Reply:
We could add some otel tracing here, in order to capture some signal. For users running locally, this would be a no-op, but for job-server and job-runner, it would emit telemetry.

If we want to avoid making the multitudinous otel libs a hard dependency on the pipeline library (and we really do want that), we could do an optimistic approach, and only do the otel stuff here if the otel libs are lazily importable.

Reply:
If we can do that without having to thread stuff up through job-server (which was my main concern) then great. We've already got one lazy import in there so what's another between friends!

Reply:
Yes, the cli does have the otel stuff, the vendoring of which is PITA

But pipeline on its own doesn't have it as a hard prod dependency currently. We can add it as a test dependency to pipeline, perhaps.

But this is starting to get to diminishing returns

Reply:
Yeah, I minded not to. Mike has already demonstrated that slow performance shows up in the existing Honeycomb traces.